### ISSUE TITLE
Instructions for building documentation seems incomplete

### ISSUE THREAD

**Issue by SSteve** [opencv/opencv#15976]
### System information (version)

- OpenCV => 4.1.2
- Operating System / Platform => macOS 10.14.6
- Compiler => Unix makefiles (Xcode 11.2)

##### Detailed description

I've started to work on cleaning up documentation (https://github.com/opencv/opencv_contrib/pull/2357, #15975) and would like to maybe add new tutorials. To this end, it would be great if I could build the documentation locally. I've tried the instructions at https://docs.opencv.org/4.1.2/d4/db1/tutorial_documentation.html, but documentation doesn't get built.

Related: #11423

##### Steps to reproduce

1. Follow the instructions at https://docs.opencv.org/4.1.2/d4/db1/tutorial_documentation.html (paying special attention to make sure `BUILD_DOCS` is enabled).
2. When executing `make doxygen`, after the **Generate Doxygen documentation** step there are multiple *unable to resolve reference* messages. Excerpt:

```
[100%] Generate Doxygen documentation
/Users/steve/Documents/GitHub/opencv/build/doc/contrib_tutorials.markdown:3: warning: unable to resolve reference to 'tutorial_table_of_content_aruco' for \ref command
/Users/steve/Documents/GitHub/opencv/build/doc/contrib_tutorials.markdown:4: warning: unable to resolve reference to 'tutorial_table_of_content_bgsegm' for \ref command
/Users/steve/Documents/GitHub/opencv/build/doc/contrib_tutorials.markdown:5: warning: unable to resolve reference to 'tutorial_table_of_content_retina' for \ref command
/Users/steve/Documents/GitHub/opencv/build/doc/contrib_tutorials.markdown:6: warning: unable to resolve reference to 'tutorial_omnidir_calib_main' for \ref command
/Users/steve/Documents/GitHub/opencv/build/doc/contrib_tutorials.markdown:7: warning: unable to resolve reference to 'tutorial_multi_camera_main' for \ref command
/Users/steve/Documents/GitHub/opencv/build/doc/contrib_tutorials.markdown:8: warning: unable to resolve reference to 'tutorial_table_of_content_cnn_3dobj' for \ref command
/Users/steve/Documents/GitHub/opencv/build/doc/contrib_tutorials.markdown:9: warning: unable to resolve reference to 'tutorial_cvv_introduction' for \ref command
/Users/steve/Documents/GitHub/opencv/build/doc/contrib_tutorials.markdown:10: warning: unable to resolve reference to 'tutorial_dnn_objdetect' for \ref command
/Users/steve/Documents/GitHub/opencv/build/doc/contrib_tutorials.markdown:11: warning: unable to resolve reference to 'tutorial_table_of_content_dnn_superres' for \ref command
/Users/steve/Documents/GitHub/opencv/build/doc/contrib_tutorials.markdown:12: warning: unable to resolve reference to 'tutorial_table_of_content_face' for \ref command
```

3. The *doc/doxygen/html/index.html* file is built, but there are no links to the documentation pages.

**Comment by alalek** [comment]
What is doxygen version? How is it installed on OSX? Did you try some Linux (Ubuntu should work)?
Please attach your `doc/Doxyfile` (perhaps as .zip archive).
Could you clean your build directory and try again?

---

I believe you can ignore these warnings. But empty `index.html` is bad.

In your PR description you can add this line to speedup Docs build:
```
force_builders_only=Docs
```

(actually this instruction will disable automatic scheduling of other builders, but this would not help a lot during "nightly" testing: 23:00 - 05:00 UTC)

Documentation builder on completion uploads documentation demo here: https://pullrequest.opencv.org/buildbot/export/pr/15975/docs (15975 is number of pull request)

**Comment by SSteve** [comment]
> What is doxygen version?

1.8.16

> How is it installed on OSX?

Homebrew

> Did you try some Linux (Ubuntu should work)?

I'll give that a try in a VM and let you know. It'll take a few hours to download here at home.

> Please attach your doc/Doxyfile (perhaps as .zip archive).

[Doxyfile.in.zip](https://github.com/opencv/opencv/files/3882191/Doxyfile.in.zip)

> Could you clean your build directory and try again?

Deleted the build directory–no change.

>In your PR description you can add this line to speedup Docs build:
>...
>Documentation builder on completion uploads documentation demo here:
>https://pullrequest.opencv.org/buildbot/export/pr/15975/docs (15975 is number of pull request)

Those are very helpful, thanks. That's a good fallback if I can't build locally.

**Comment by alalek** [comment]
> Doxyfile.in.zip

I mean file from the "build" directory. `.in` file is from the source tree.

> WIP

Yes.

**Comment by SSteve** [comment]
Oops. 
[Doxyfile.zip](https://github.com/opencv/opencv/files/3882658/Doxyfile.zip)

**Comment by SSteve** [comment]
Under Ubuntu 16.04 I get these warnings:

```
/media/psf/Home/Development/GitHub/opencv/doc/tutorials/introduction/documenting_opencv/documentation_tutorial.markdown:508: warning: Found unknown command `\add_toggle_cpp'
/media/psf/Home/Development/GitHub/opencv/doc/tutorials/introduction/documenting_opencv/documentation_tutorial.markdown:513: warning: Found unknown command `\end_toggle'
/media/psf/Home/Development/GitHub/opencv/doc/tutorials/introduction/documenting_opencv/documentation_tutorial.markdown:515: warning: Found unknown command `\add_toggle_java'
/media/psf/Home/Development/GitHub/opencv/doc/tutorials/introduction/documenting_opencv/documentation_tutorial.markdown:520: warning: Found unknown command `\end_toggle'
/media/psf/Home/Development/GitHub/opencv/doc/tutorials/introduction/documenting_opencv/documentation_tutorial.markdown:522: warning: Found unknown command `\add_toggle_python'
/media/psf/Home/Development/GitHub/opencv/doc/tutorials/introduction/documenting_opencv/documentation_tutorial.markdown:527: warning: Found unknown command `\end_toggle'
```

The index.html file is successfully created with links. In spite of the warnings, the code toggle buttons in documentation_tutorial work correctly. So, for me, this is a workable solution to this issue. I'll let you decide whether to leave this issue open for building documentation on macOS (or to close this one and create one specifically for macOS).

**Comment by alalek** [comment]
It looks strange (we using Ubuntu to generate documentation).

Could you try to use build directory outside of source tree?

```.diff
-/Users/steve/Development/GitHub/opencv/build
+/Users/steve/Development/GitHub/build_opencv
```

**Comment by SSteve** [comment]
I deleted the `build` directory under `opencv` and created a new build directory outside the source tree. I get the same results. Here's an excerpt of the warnings:

```
[100%] Generate Doxygen documentation
/Users/steve/Development/GitHub/opencv-macos-build/doc/contrib_tutorials.markdown:3: warning: unable to resolve reference to 'tutorial_table_of_content_aruco' for \ref command
/Users/steve/Development/GitHub/opencv-macos-build/doc/contrib_tutorials.markdown:4: warning: unable to resolve reference to 'tutorial_table_of_content_bgsegm' for \ref command
/Users/steve/Development/GitHub/opencv-macos-build/doc/contrib_tutorials.markdown:5: warning: unable to resolve reference to 'tutorial_table_of_content_retina' for \ref command
/Users/steve/Development/GitHub/opencv-macos-build/doc/contrib_tutorials.markdown:6: warning: unable to resolve reference to 'tutorial_multi_camera_main' for \ref command
/Users/steve/Development/GitHub/opencv-macos-build/doc/contrib_tutorials.markdown:7: warning: unable to resolve reference to 'tutorial_omnidir_calib_main' for \ref command
/Users/steve/Development/GitHub/opencv-macos-build/doc/contrib_tutorials.markdown:8: warning: unable to resolve reference to 'tutorial_table_of_content_cnn_3dobj' for \ref command
```

**Comment by mshabunin** [comment]
@SSteve , doxygen 1.8.16 seem to be broken, please stick to doxygen 1.8.12 or 1.8.14 for now.

**Comment by mshabunin** [comment]
Issue: https://github.com/doxygen/doxygen/issues/7190 have been fixed but not released yet.

**Comment by SSteve** [comment]
Reverting to doxygen 1.8.14 fixed the issue under macOS 10.14.6.

Also, the warnings I got with Ubuntu 16.04 went away when I upgraded to Ubuntu 18.04.
